package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+    cdef real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef bint prepared
+    cdef NSGrid grid
+
+    def __init__(self, u, cutoff, coords=None, prepare=True, debug=False, max_gridsize=5000):
+        import MDAnalysis as mda
+        from MDAnalysis.lib.mdamath import triclinic_vectors
+
+        self.debug = debug
+
+        if not isinstance(u, mda.Universe):
+            raise TypeError("FastNS class must be initialized with a valid MDAnalysis.Universe instance")
+        box = triclinic_vectors(u.dimensions)
+
+        self.box = PBCBox(box)
+
+
+        if coords is None:
+            coords = u.atoms.positions
+
+        self.coords = coords.copy()
+
+        self.coords_bbox =  self.box.fast_put_atoms_in_bbox(coords)
+
+        if cutoff < 0:
+            raise ValueError("Cutoff must be positive!")
+        if cutoff * cutoff > self.box.c_pbcbox.max_cutoff2:
+            raise ValueError("Cutoff greater than maximum cutoff ({:.3f}) given the PBC")
+        self.cutoff = cutoff
+
+        self.grid = NSGrid(self.coords_bbox.shape[0], cutoff, self.box, max_gridsize, debug=debug)
+        self.prepared = False
+        if prepare:
+            self.prepare()
+
+
+    def prepare(self, force=False):
+        if self.prepared and not force:
+            return
+
+        self.grid.fill_grid(self.coords_bbox)
+
+        self.prepared = True
+
+    def search(self, search_ids=None):
+        cdef ns_int i, j, size_search
+        cdef ns_int d, m
+        cdef NSResults results
+        cdef ns_int size = self.coords_bbox.shape[0]
+
+        cdef ns_int current_beadid, bid
+        cdef rvec current_coords
+
+        cdef ns_int cellindex, cellindex_adjacent, cellindex_probe
+        cdef ivec cellxyz, debug_cellxyz
+
+        cdef real[:, ::1] search_coords
+        cdef ns_int[:] search_ids_view
+
+        cdef ns_int xi, yi, zi
+        cdef real d2
+        cdef rvec shifted_coord, probe, dx
+
+        cdef ns_int nchecked = 0
+
+        cdef real cutoff2 = self.cutoff * self.cutoff
+        cdef ns_int[:] checked
+        cdef ns_int npairs = 0
+
+        #cdef bint debug=False
+
+
+        if not self.prepared:
+            self.prepare()
+
+        if search_ids is None:
+            search_ids=np.arange(size)
+        elif type(search_ids) == np.int:
+            search_ids = np.array([search_ids,], dtype=np.int)
+        elif type(search_ids) != np.ndarray:
+            search_ids = np.array(search_ids, dtype=np.int)
+
+        search_ids_view = search_ids
+        size_search = search_ids.shape[0]
+
+        checked = np.zeros(size, dtype=np.int)
+
+        results = NSResults(self.cutoff, self.coords, search_ids, self.debug)
+
+        cdef bint memory_error = False
+
+        # if self.debug and debug:
+        #     print("FastNS: Debug flag is set to True for FastNS.search()")
+
+        with nogil:
+            for i in range(size_search):
+                if memory_error:

```

Comment:
where is this value ever changed? I see all the checks and they sound nice but I don't see where you set this value

Reply:
What do you mean? when memory error is set to True? it is if the `NSResults.resize`fails to reallocate memory.

Reply:
I found it later. It's a bit hidden away in the nested loops.

Reply:
Yes, unfortunately, I do not see any cleaner way to propagate the memory allocation failure without requiring the GIL (and thus making the code slower)